lib/jsdom/living/helpers/http-request.js

```diff
@@ -0,0 +1,272 @@
+    const jar = self._options.cookieJar;
+    const cookies = res.headers["set-cookie"];
+    if (!this._areCookiesDisabled && jar && typeof jar.setCookieSync === "function" && Array.isArray(cookies)) {
+      try {
+        cookies.forEach(cookie => jar.setCookieSync(cookie, self._currentUrl, { ignoreError: true }));
+      } catch (e) {
+        this.emit("error", e);
+      }
+    }
+  }
+
+  write(data, encoding) {
+    if (this._ending) {
+      const writeAfterEndError = new Error("Writing data after ClientRequest.end() is called is not allowed");
+      writeAfterEndError.code = "ERR_STREAM_WRITE_AFTER_END";
+      throw writeAfterEndError;
+    }
+    if (!(typeof data === "string" || (typeof data === "object" && ("length" in data)))) {
+      throw new TypeError("data should be a string, Buffer or Uint8Array");
+    }
+    if (data.length > 0) {
+      this._requestBodyLength += data.length;
+      this._requestBodyBuffers.push({ data, encoding });
+      this._currentRequest.write(data, encoding);
+    }
+  }
+
+  end() {
+    this.emit("request", this._currentRequest);
+    this._ended = this._ending = true;
+    this._currentRequest.end();
+  }
+
+  setHeader(name, value) {
+    this._options.headers[name] = value;
+    this._currentRequest.setHeader(name, value);
+  }
+
+  removeHeader(name) {
+    delete this._options.headers[name];
+    this._currentRequest.removeHeader(name);
+  }
+
+  toJSON() {
+    const uri = new URL(this._currentUrl);
+    const { method, headers } = this._options;
+    return { uri, method, headers };
+  }
+
+  _performRequest() {
+    const urlOptions = new URL(this._currentUrl);
+    const { protocol: scheme } = urlOptions;
+    this._options.agent = this._options.agents[scheme.substr(0, scheme.length - 1)];
+    const self = this;
+    function onNative(response) {
+      self._processResponse(response);
+    }
+    const { request } = scheme === "https:" ? https : http;
+    const message = request(this._currentUrl, this._options, onNative);
+    this._currentRequest = message;
+    this.getCookie();
+    ["connect", "error", "socket", "timeout"].forEach(event => {
+      message.on(event, (...args) => {
+        self.emit(event, ...args);
+      });
+    });
+    if (this._isRedirect) {
+      let i = 0;
+      const buffers = this._requestBodyBuffers;
+      (function writeNext(error) {
+        if (message === self._currentRequest) {
+          if (error) {
+            self.emit("error", error);
+          } else if (i < buffers.length) {
+            const buffer = buffers[i++];
+            if (!message.writableEnded) {
+              message.write(buffer.data, buffer.encoding, writeNext);
+            }
+          } else if (self._ended) {
+            message.end();
+          }
+        }
+      })();
+    }
+  }
+
+  _processResponse(response) {
+    response.request = {};
+    response.request.headers = this._options.headers;
+    response.request.uri = new URL(this._currentUrl);
```

Comment:
uri => url

Reply:
The `property `response.request.uri` is accessed by XMLHTTPRequestImpl.js in its redirect event listener, to mention one place.

```javascript
      client.on("redirect", () => {
        const { response } = client;
        const destUrlObj = new URL(response.request.headers.Referer);

        const urlObj = new URL(response.request.uri.href);

        if (destUrlObj.origin !== urlObj.origin && destUrlObj.origin !== flag.origin) {
          properties.origin = "null";
        }

        response.request.headers.Origin = properties.origin;

        if (flag.origin !== destUrlObj.origin &&
            destUrlObj.protocol !== "data:") {
          if (!xhrUtils.validCORSHeaders(this, response, flag, properties, flag.origin)) {
            return;
          }
          if (urlObj.username || urlObj.password) {
            properties.error = "Userinfo forbidden in cors redirect";
            xhrUtils.dispatchError(this);
          }
        }
      });
```